### PR TITLE
Add geometry_union scalar function to efficiently union an array of geometries

### PR DESCRIPTION
**Pull request by tdcmeehan**
A benchmark has been added to measure geometry aggregation performance.
Here are some benchmark numbers which demonstrate the performance over
50 states:

```
Benchmark                                         Mode  Cnt    Score    Error  Units
BenchmarkGeometryAggregation.benchmarkArrayUnion  avgt   30   24.601 ±  0.659  ms/op
BenchmarkGeometryAggregation.benchmarkConvexHull  avgt   30   23.169 ±  0.673  ms/op
BenchmarkGeometryAggregation.benchmarkUnion       avgt   30  647.646 ± 14.411  ms/op

### CODE REVIEW

**Comment by tdcmeehan** [comment]
Fixes #11420

File: presto-docs/src/main/sphinx/functions/geospatial.rst [hunk 1]
```diff
@@ -144,7 +144,11 @@ Operations
 
     Returns a geometry that represents the point set union of the input geometries.
 
-    This function doesn't support geometry collections.
+.. function:: ST_Union(array<Geometry>) -> Geometry
```
Comment (mbasmanova) [inline presto-docs/src/main/sphinx/functions/geospatial.rst]: @tdcmeehan Tim,

SQL/MM spec defines `ST_Geometries(array<Geometry>)` function that constructs a geometry collection from an array of geometries. Would it make sense to add this function and use existing `ST_Union(Geometry)`? I think `ST_Geometries` is similar to PostGIS's `ST_Collect`.

screen shot 2018-10-02 at 8 14 07 am

Comment (mbasmanova) [reply presto-docs/src/main/sphinx/functions/geospatial.rst]: @tdcmeehan Tim, today `ST_Union(ST_Geometries(ARRAY[...]))` would be less efficient than `ST_Union(ARRAY[...])` because of extra cost of serialization and deserialization of geometry returned by `ST_Geometries`. However, we discussed at some point to allow `Block` to store any object, not just `Slice`. Once we do that, `ST_Union(ST_Geometries(ARRAY[...]))` will be as efficient as `ST_Union(ARRAY[...])`. CC: @martint

Comment (tdcmeehan) [reply presto-docs/src/main/sphinx/functions/geospatial.rst]: There are two components to this feedback, which have different impact on performance.

The first is to add `ST_Geometries` and have union operate on that.  I quickly wrote this up, for my JMH benchmark, I don't see any performance impact outside the margin of error:

```
Benchmark                                              Mode  Cnt   Score   Error  Units
BenchmarkGeometryAggregation.benchmarkGeometriesUnion  avgt   30  25.195 ± 0.584  ms/op
```

The second is to reuse the existing implementation of `ST_Union`.  What I presume you mean is overload `ST_Union` to accept just a single geometry, since the current implementation takes in two.  If I do that, and have it do something like take the first element of a geometry collection and union it with the entire geometry collection (reusing the existing `ST_Union` code), it's almost 50% slower.

```
Benchmark                                              Mode  Cnt   Score   Error  Units
BenchmarkGeometryAggregation.benchmarkGeometriesUnion  avgt   30  35.341 ± 1.170  ms/op
```

I suspect the reason for this is whenever the library detects a heterogenous union (either one or the other sides is a geometry collection, or the types of either side of the union don't match), it chooses a more robust algorithm to union, but which is slower.

Since there isn't much observed overhead by adding the `ST_Geometries`, what if I remove the public addition of the array form of `ST_Union`, but keep my private implementation in as shown in this PR, and add `ST_Geometries` so that users can still get this speedup, without us adding non-standard array form of `ST_Union`?

Comment (mbasmanova) [reply presto-docs/src/main/sphinx/functions/geospatial.rst]: @tdcmeehan Tim, sorry, I got confused. I was thinking of `ST_ConvexHull` which takes a single geometry and somehow forgetting that `ST_Union` takes two. I think `ST_Union` taking an array of geometries like you proposed initially makes sense. It is not part of the standard, but it is part of PostGIS and it makes sense generally. `ST_Union` taking a single geometry also not part of the standard, but is it not part of PostGIS either.

Let's go with `ST_Union(array<Geometry>)`. I'll review the changes.

**Comment by dain** [comment]
I added the `syntax-needs-review` tag to this for consistency, but it should be easy to clear this one as the `ST_*` functions are covered by the spec.

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 2]
```diff
@@ -1242,4 +1291,28 @@ private static boolean envelopes(Slice left, Slice right, EnvelopesPredicate pre
     {
         boolean apply(Envelope left, Envelope right);
     }
+
+    private static Iterable<Slice> getSlicesFromBlock(Type type, Block block)
+    {
+        requireNonNull(type, "type is null");
+        requireNonNull(block, "block is null");
+        return () -> new Iterator<Slice>() {
+            private int blockPositionCount = block.getPositionCount();
```
Comment (mbasmanova) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: final

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 3]
```diff
@@ -1242,4 +1291,28 @@ private static boolean envelopes(Slice left, Slice right, EnvelopesPredicate pre
     {
         boolean apply(Envelope left, Envelope right);
     }
+
+    private static Iterable<Slice> getSlicesFromBlock(Type type, Block block)
```
Comment (mbasmanova) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: since this function only used for geometries, perhaps, simplify to `getGeometrySlicesFromBlock(block)`

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/BenchmarkGeometryAggregation.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static org.openjdk.jmh.annotations.Mode.AverageTime;
+import static org.openjdk.jmh.annotations.Scope.Thread;
+
+@OutputTimeUnit(MILLISECONDS)
+@BenchmarkMode(AverageTime)
+@Fork(3)
+@Warmup(iterations = 10)
+@Measurement(iterations = 10)
+public class BenchmarkGeometryAggregation
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/BenchmarkGeometryAggregation.java]: s/BenchmarkGeometryAggregation/BenchmarkGeometryAggregations

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/BenchmarkGeometryAggregation.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+            queryRunner.createCatalog("memory", new MemoryConnectorFactory(), ImmutableMap.of());
+
+            Path path = Paths.get(BenchmarkGeometryAggregation.class.getClassLoader().getResource("us-states.tsv").getPath());
+            String polygonValues = Files.lines(path)
+                    .map(line -> line.split("\t"))
+                    .map(parts -> format("('%s', '%s')", parts[0], parts[1]))
+                    .collect(Collectors.joining(","));
+
+            queryRunner.execute(
+                    format("CREATE TABLE memory.default.us_states AS SELECT st_geometryfromtext(t.wkt) as geom FROM (VALUES %s) as t (name, wkt)",
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/BenchmarkGeometryAggregation.java]: s/st_geometryfromtext/ST_GeometryFromText

s/as/AS

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/BenchmarkGeometryAggregation.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+            queryRunner.close();
+            queryRunner = null;
+        }
+    }
+
+    @Benchmark
+    public MaterializedResult benchmarkArrayUnion(Context context)
+    {
+        return context.getQueryRunner()
+                .execute("SELECT ST_Union(array_agg(p.geom)) FROM us_states p ");
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/BenchmarkGeometryAggregation.java]: nit: extra space after `p`

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/BenchmarkGeometryAggregation.java [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    @Benchmark
+    public MaterializedResult benchmarkUnion(Context context)
+    {
+        return context.getQueryRunner()
+                .execute("SELECT geometry_union_agg(p.geom) FROM us_states p ");
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/BenchmarkGeometryAggregation.java]: nit: extra space after `p`

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/BenchmarkGeometryAggregation.java [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    @Benchmark
+    public MaterializedResult benchmarkConvexHull(Context context)
+    {
+        return context.getQueryRunner()
+                .execute("SELECT convex_hull_agg(p.geom) FROM us_states p ");
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/BenchmarkGeometryAggregation.java]: nit: extra space after `p`

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/aggregation/TestGeometryUnionGeoAggregation.java [hunk 9]
```diff
@@ -13,12 +13,23 @@
...
+import java.util.List;
+
+import static com.facebook.presto.plugin.geospatial.GeometryType.GEOMETRY;
+import static java.lang.String.format;
+import static java.util.stream.Collectors.toList;
+
 public class TestGeometryUnionGeoAggregation
         extends AbstractTestGeoAggregationFunctions
 {
+    public static final Joiner COMMA_JOINER = Joiner.on(",");
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/aggregation/TestGeometryUnionGeoAggregation.java]: private

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/aggregation/TestGeometryUnionGeoAggregation.java [hunk 10]
```diff
@@ -292,47 +298,65 @@
...
 
     @Override
     protected String getFunctionName()
     {
         return "geometry_union_agg";
     }
+
+    private void assertArrayAggAndStUnion(String testDescription, String expectedWkt, String[] wkts)
+    {
+        List<String> wktList = Arrays.stream(wkts).map(wkt -> format("ST_GeometryFromText('%s')", wkt != null && !wkt.trim().isEmpty() ? wkt : "")).collect(toList());
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/aggregation/TestGeometryUnionGeoAggregation.java]: let's drop `!wkt.trim().isEmpty()` for simplicity

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/aggregation/TestGeometryUnionGeoAggregation.java [hunk 11]
```diff
@@ -292,47 +298,65 @@
...
+        String wktArray = format("ARRAY[%s]", COMMA_JOINER.join(wktList));
+        // ST_Union(ARRAY[ST_GeometryFromText('...'), ...])
+        assertFunction(format("ST_Union(%s)", wktArray), GEOMETRY, expectedWkt);
+        Collections.reverse(wktList);
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/aggregation/TestGeometryUnionGeoAggregation.java]: nit: perhaps, add an empty line to visually separate two test cases

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 12]
```diff
@@ -561,27 +573,64 @@ public static long stNumGeometries(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
         return ((OGCGeometryCollection) geometry).numGeometries();
     }
 
-    @Description("Returns a geometry that represents the point set union of the input geometries. This function doesn't support geometry collections.")
+    @Description("Returns a geometry that represents the point set union of the input geometries.")
```
Comment (mbasmanova) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: Thanks for removing this restriction. Would you move that change into a separate commit?

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java [hunk 13]
```diff
@@ -160,4 +170,53 @@ public static boolean isPointOrRectangle(OGCGeometry ogcGeometry, Envelope envel
 
         return true;
     }
+
+    public static Iterable<OGCGeometry> flattenCollection(OGCGeometry geometry)
+    {
+        if (geometry == null) {
+            return emptyList();
```
Comment (mbasmanova) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java]: ImmutableList.of()

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java [hunk 14]
```diff
@@ -160,4 +170,53 @@ public static boolean isPointOrRectangle(OGCGeometry ogcGeometry, Envelope envel
...
+        }
+        if (!(geometry instanceof OGCConcreteGeometryCollection)) {
+            return singleton(geometry);
```
Comment (mbasmanova) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java]: ImmutableList.of(geometry)

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java [hunk 15]
```diff
@@ -160,4 +170,53 @@ public static boolean isPointOrRectangle(OGCGeometry ogcGeometry, Envelope envel
...
+        }
+        if (((OGCConcreteGeometryCollection) geometry).numGeometries() == 0) {
+            return emptyList();
```
Comment (mbasmanova) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java]: ImmutableList.of()

**Comment by dain** [comment]
> > I added the syntax-needs-review tag to this for consistency, but it should be easy to clear this one as the ST_* functions are covered by the spec.
> 
> @dain What's the process for clearing this tag? The proposed variant of the `ST_Union` function is not covered by the spec, but it makes sense in general.

Normally, we just defer to @martint.  In this case, I'm concerned about any function starting with `ST_` that is not covered by the spec.  The spec has clearly carved out this space for their functions, and I think we should leave it for them to make any future changes.

**Comment by tdcmeehan** [comment]
Could a single-argument form of `ST_Union` be considered the same as this (5.1.19 in SQL/MM part 3):

> 2) The null-call method ST_Union(ST_Geometry) returns an ST_Geometry value that represents the point set union: Closure(SELF ∪ ageometry).

If so, I would go the route of implementing `ST_Geometries` and feeding it to this.

**Comment by tdcmeehan** [comment]
@martint and I discussed, how about `geometry_union` (continuing usage of `geometry_` namespace distinct from `ST_` namespace)?

**Comment by mbasmanova** [comment]
> how about geometry_union (continuing usage of geometry_ namespace distinct from ST_ namespace)?

That should work, but it is confusing to our users who don't quite appreciate the reasons behind usage of a different name.

**Comment by tdcmeehan** [comment]
@mbasmanova perhaps we separate out `ST_` functions, which presume perfect compliance to SQL/MM part 3, and custom functions which begin with `geometry_`, `bing_`, and highlight in the documentation that these are extensions which may be helpful (and perhaps even refer to analogs in other popular geospatial databases).  While I originally named this function because it existed in PostGIS, rather than coding to PostGIS perhaps we should code to the standard, and clearly highlight any non-standard functions.

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java [hunk 16]
```diff
@@ -160,4 +169,53 @@ public static boolean isPointOrRectangle(OGCGeometry ogcGeometry, Envelope envel
 
         return true;
     }
+
+    public static Iterable<OGCGeometry> flattenCollection(OGCGeometry geometry)
```
Comment (mbasmanova) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java]: Since this function is used only once, I'd move it to `GeoFunctions.java`.

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/aggregation/TestGeometryUnionGeoAggregation.java [hunk 17]
```diff
@@ -292,47 +298,66 @@
...
+        assertArrayAggAndStUnion(testDescription, expectedWkt, wkts);
     }
 
     @Override
     protected String getFunctionName()
     {
         return "geometry_union_agg";
     }
+
+    private void assertArrayAggAndStUnion(String testDescription, String expectedWkt, String[] wkts)
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/aggregation/TestGeometryUnionGeoAggregation.java]: the implementation doesn't use `ST_Union` function, hence, the name is misleading; how about `assertGeometryUnion`?

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 18]
```diff
@@ -571,27 +572,12 @@ public static long stNumGeometries(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
...
```
Comment (mbasmanova) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: Docs need to be updated to remove `This function doesn’t support geometry collections.` from `ST_Union` as well.

Comment (tdcmeehan) [reply presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: I did that, but it got squished into the other commit.  I'll fix it so it's in the right commit.

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 19]
```diff
@@ -561,27 +573,64 @@ public static long stNumGeometries(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
...
+        List<OGCGeometry> outputs = new ArrayList<>();
+        for (GeometryCursor operator : operatorsByDimension) {
+            OGCGeometry unionedGeometry = createFromEsriGeometry(operator.next(), null);
+            if (unionedGeometry != null) {
+                outputs.add(unionedGeometry);
+            }
         }
 
-        return serialize(leftGeometry.union(rightGeometry));
+        if (outputs.isEmpty()) {
```
Comment (martint) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: Does this happen when all geometries are empty? If so, this appears to be a change in semantics. The old version of the code seemed to do:

```
if (rightGeometry.isEmpty()) {
    return left;
}
```

**Comment by tdcmeehan** [comment]
@martint I made a slight change so `GEOMETRYCOLLECTION EMPTY` is preserved if the inputs consisted solely of empty collections (second bullet above).  It's squashed, but here is the commit of that diff: 524df0b97f4

File: presto-docs/src/main/sphinx/functions/geospatial.rst [hunk 20]
```diff
@@ -106,6 +106,12 @@ Relationship Tests
 Operations
 ----------
 
+.. function:: geometry_union(array(Geometry)) -> Geometry
+
+    Returns a geometry that represents the point set union of the input geometries.  Performance
```
Comment (nezihyigitbasi) [inline presto-docs/src/main/sphinx/functions/geospatial.rst]: nit: multiple spaces before `  Performance`

File: presto-docs/src/main/sphinx/functions/geospatial.rst [hunk 21]
```diff
@@ -146,6 +152,8 @@ Operations
 
     This function doesn't support geometry collections.
 
+    See also:  :func:`geometry_union`, :func:`geometry_union_agg`
```
Comment (nezihyigitbasi) [inline presto-docs/src/main/sphinx/functions/geospatial.rst]: Not a sphinx expert, but `see also` are separate sections in other doc/rst files.

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 22]
```diff
@@ -707,7 +772,7 @@ public static Slice stStartPoint(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
     @ScalarFunction("simplify_geometry")
     @SqlType(GEOMETRY_TYPE_NAME)
     public static Slice simplifyGeometry(@SqlType(GEOMETRY_TYPE_NAME) Slice input,
-                                         @SqlType(DOUBLE) double distanceTolerance)
```
Comment (nezihyigitbasi) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: whitespace change

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 23]
```diff
@@ -1242,4 +1307,76 @@ private static boolean envelopes(Slice left, Slice right, EnvelopesPredicate pre
...
         boolean apply(Envelope left, Envelope right);
     }
+
+    private static Iterable<Slice> getGeometrySlicesFromBlock(Block block)
+    {
+        requireNonNull(block, "block is null");
+        return () -> new Iterator<Slice>() {
+            private final int blockPositionCount = block.getPositionCount();
+            private int iteratorPosition;
+            @Override
```
Comment (nezihyigitbasi) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: newline before `Override`

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 24]
```diff
@@ -1242,4 +1307,76 @@ private static boolean envelopes(Slice left, Slice right, EnvelopesPredicate pre
...
+            public boolean hasNext()
+            {
+                return iteratorPosition != blockPositionCount;
```
Comment (nezihyigitbasi) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: Why not call `block.getPositionCount()` here instead of having a `final` variable? Does it make any diff in perf?

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeometryType.java [hunk 25]
```diff
@@ -59,12 +59,20 @@ public Slice getSlice(Block block, int position)
     @Override
     public void writeSlice(BlockBuilder blockBuilder, Slice value)
     {
+        if (value == null) {
```
Comment (nezihyigitbasi) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeometryType.java]: Is this a bug that you found while adding this new function? If it is, it would be nice to fix this in a separate commit to make it easier to track.
